util/omrutil/unix/linux/riscv/32/cas32helper.s

```diff
@@ -0,0 +1,57 @@
+###############################################################################
+# Copyright (c) 2019, 2019 IBM Corp. and others
+#
+# This program and the accompanying materials are made available under
+# the terms of the Eclipse Public License 2.0 which accompanies this
+# distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+# or the Apache License, Version 2.0 which accompanies this distribution and
+# is available at https://www.apache.org/licenses/LICENSE-2.0.
+#
+# This Source Code may also be made available under the following
+# Secondary Licenses when the conditions for such availability set
+# forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+# General Public License, version 2 with the GNU Classpath
+# Exception [1] and GNU General Public License, version 2 with the
+# OpenJDK Assembly Exception [2].
+#
+# [1] https://www.gnu.org/software/classpath/license.html
+# [2] http://openjdk.java.net/legal/assembly-exception.html
+#
+# SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+###############################################################################
+
+	.file	"cas32helper.s"
+	.option nopic
+	.text
+
+# Prototype: uint32_t RiscvCAS32Helper(volatile uint32_t *addr, uint32_t compareValue, uint32_t swapValue);
+# Defined in: #Args: 3
+	.align	2
+	.globl	RiscvCAS32Helper
+	.type	RiscvCAS32Helper, @function
+RiscvCAS32Helper:
+	addi	sp,sp,-32
+	sd	s0,24(sp)
+	addi	s0,sp,32
```

Comment:
Why doesn't the code below access arguments via `sp` (with offsets increased by 32) so we can avoid saving and restoring `s0`? Even better, don't modify `sp` at all (this is a leaf function). The same applies to `cas64helper.s`.

Reply:
All the assembly code were added against the disassembled C function on RISC-V. 
```
       addi	 sp,sp,-32
	sd	s0,24(sp)
	addi	s0,sp,32
...
	ld	s0,24(sp)
	addi	sp,sp,32
	jr	ra
```
In theory, it might work by changing the stack frame operations before entering/existing the function, but I didn't deeply dive into the details of how `s0` and `sp`works in saving/restoring the function stack and what else registers change with these operations (given there is no explanation on the Spec /it is totally based on the experience of programmers and the compilation result). So we are trying to avoid changing the existing invoking pattern in functions on RISC-V in case of unexpected behaviour which is difficult to investigate in the future.

Reply:
RISC-V ABI is can be found here: https://riscv.org/wp-content/uploads/2015/01/riscv-calling.pdf
You can use `t0` instead of `s0` then you need not to restore it (`tX` are caller-saved, `sX` are callee-saved). 
Leaving `sp` intact and adjusting offsets accordingly should work fine.

Reply:
I already checked riscv-calling.pdf previously and it might work, but I am wondering whether we really need to optimize in this way given all C functions on RISC-V are compiled by using `s0 `and `sp` to deal with the the stack operations.

Reply:
> I already checked riscv-calling.pdf previously and it might work
Ah, ok. I somehow thought you may have missed this one.

> whether we really need to optimize in this way given all C functions on RISC-V are compiled...
Yep, I agree. If we need to optimize this, we may want to make it an inline assembly - this would give the compiler a chance to inline and eliminate all the prolog/epilogue altogether.

Reply:
@keithc-ca, I think we should leave the code unchanged there given that there is no noticeable difference in term performance without JIT even though this assembly is optimized.

In addition, I am currently investigating the crash with `java -version` on the hardware (not sure how long to get this issue resolved as it is specific to the Fedora). Optimizing any assembly like this might incur unexpected behaviour on the hardware if incorrect, which gets things even more complicated during the analysis on the crash.